# EDA, Descriptive Statistics and Model Developmet of AutoMobile Dataset.

#### _Import all the necessary packages_

In [2]:
import pandas as pd

In [16]:
raw_data = pd.read_csv("Automobile_data.csv")

In [ ]:
raw_data.shape
# rows X columns

In [17]:
raw_data.describe(include="all")
#This will give you all the basic statistic measurements of all the columns in the dataset

symboling normalized-losses    make fuel-type aspiration  \
count   205.000000               205     205       205        205   
unique         NaN                52      22         2          2   
top            NaN                 ?  toyota       gas        std   
freq           NaN                41      32       185        168   
mean      0.834146               NaN     NaN       NaN        NaN   
std       1.245307               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           205        205          205             205  205.000000  ...   
unique            3          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            114         96          120             202         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.756585  ...   
std             NaN        NaN          NaN             NaN    6.021776  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    205.000000          205   205     205        205.000000        205   
unique          NaN            8    39      37               NaN         60   
top             NaN         mpfi  3.62     3.4               NaN         68   
freq            NaN           94    23      20               NaN         19   
mean     126.907317          NaN   NaN     NaN         10.142537        NaN   
std       41.642693          NaN   NaN     NaN          3.972040        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        205  205.000000  205.000000   205  
unique        24         NaN         NaN   187  
top         5500         NaN         NaN     ?  
freq          37         NaN         NaN     4  
mean         NaN   25.219512   30.751220   NaN  
std          NaN    6.542142    6.886443   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

In [ ]:
raw_data.head()
#This will display first 5 rows of the entire dataset.

#### The first observation from the dataset is that there are "?" entries. Hence, you can either remove the entire row containig the "?" values or replace them with np.nan so that we can analyse the dataset later

In [18]:
#Let's replace the "?" with NaN values.

raw_data.replace("?",np.nan,inplace = True)

In [ ]:
raw_data.dtypes

### Let's check for missing values

In [19]:
raw_data.isna().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

In [20]:
#Let's calculate the percentage of missing values in each column.
perc_missing = pd.DataFrame((raw_data.isna().sum()/len(raw_data)) * 100,columns = ["Perecentage Missing"])

In [21]:
perc_missing

Perecentage Missing
symboling                      0.00000
normalized-losses             20.00000
make                           0.00000
fuel-type                      0.00000
aspiration                     0.00000
num-of-doors                   0.97561
body-style                     0.00000
drive-wheels                   0.00000
engine-location                0.00000
wheel-base                     0.00000
length                         0.00000
width                          0.00000
height                         0.00000
curb-weight                    0.00000
engine-type                    0.00000
num-of-cylinders               0.00000
engine-size                    0.00000
fuel-system                    0.00000
bore                           1.95122
stroke                         1.95122
compression-ratio              0.00000
horsepower                     0.97561
peak-rpm                       0.97561
city-mpg                       0.00000
highway-mpg                    0.00000
price                          1.95122

# Based on the above observations, each column has 205 rows of data, with the following columns containing missing data:

1. "normalized-losses": 41 missing data
2. "num-of-doors": 2 missing data
3. "bore": 4 missing data
4. "stroke" : 4 missing data
5. "horsepower": 2 missing data
6. "peak-rpm": 2 missing data
7. "price": 4 missing data

### How can we deal with missing data

1. Drop data 
    1. drop the whole row
    2. drop the whole column
2. Replace data
    1. replace it by mean
    2. replace it by frequency
    3. replace it based on other functions (Fit a model with missing value column as target variable and predict with the          help of other values

In [ ]:
#Let's replace the missing values with the mean of the column

In [11]:
raw_data.mean()

symboling               0.834146
wheel-base             98.756585
length                174.049268
width                  65.907805
height                 53.724878
curb-weight          2555.565854
engine-size           126.907317
compression-ratio      10.142537
city-mpg               25.219512
highway-mpg            30.751220
dtype: float64

In [22]:
cols = ["normalized-losses","bore","stroke","horsepower","peak-rpm","price"]

In [23]:
for i in cols:
    avg = raw_data[i].astype(float).mean(axis=0)
    raw_data[i].replace(np.nan,avg,inplace=True)
    print(avg)

122.0
3.3297512437810957
3.2554228855721337
104.25615763546799
5125.369458128079
13207.129353233831


In [24]:
raw_data["num-of-doors"].mode()

0    four
dtype: object

In [25]:
raw_data["num-of-doors"].replace(np.nan,"four",inplace=True)

In [27]:
raw_data.isna().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

In [47]:
data=raw_data

In [48]:
pd.set_option('display.max_columns', None)

In [49]:
#Converting datatypes of certain columns to fit the datatpes of values in it.
data.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

In [40]:
data.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               122  alfa-romero       gas        std          two   
1          3               122  alfa-romero       gas        std          two   
2          1               122  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  length  width  \
0  convertible          rwd           front        88.6   168.8   64.1   
1  convertible          rwd           front        88.6   168.8   64.1   
2    hatchback          rwd           front        94.5   171.2   65.5   
3        sedan          fwd           front        99.8   176.6   66.2   
4        sedan          4wd           front        99.4   176.6   66.4   

   height  curb-weight engine-type num-of-cylinders  engine-size fuel-system  \
0    48.8         2548        dohc             four          130        mpfi   
1    48.8         2548        dohc             four          130        mpfi   
2    52.4         2823        ohcv              six          152        mpfi   
3    54.3         2337         ohc             four          109        mpfi   
4    54.3         2824         ohc             five          136        mpfi   

   bore stroke  compression-ratio horsepower peak-rpm  city-mpg  highway-mpg  \
0  3.47   2.68                9.0        111     5000        21           27   
1  3.47   2.68                9.0        111     5000        21           27   
2  2.68   3.47                9.0        154     5000        19           26   
3  3.19    3.4               10.0        102     5500        24           30   
4  3.19    3.4                8.0        115     5500        18           22   

   price  
0  13495  
1  16500  
2  16500  
3  13950  
4  17450

In [52]:
data[["normalized-losses","bore","stroke","horsepower","peak-rpm","price"]]=data[["normalized-losses","bore","stroke","horsepower","peak-rpm","price"]].astype(float)